src/ra_wm_handler.cpp

```diff
@@ -0,0 +1,227 @@
+    // since we can't assume what a window actually has
+
+    // This can be used for Modals too but not in WM_CLOSE
+    if (hwnd != nullptr)
+    {
+        ::DestroyWindow(hwnd);
+        hwnd = nullptr;
+    }
+}
+
+_Use_decl_annotations_
+LRESULT WM_Handler::OnNotify(HWND hwnd, int idFrom, NMHDR* pnmhdr) noexcept
+{
+    return FORWARD_WM_NOTIFY(hwnd, idFrom, pnmhdr, SendMessage);
+}
+
+_Use_decl_annotations_
+void WM_Handler::OnTimer(HWND hwnd, UINT id) noexcept { FORWARD_WM_TIMER(hwnd, id, PostMessage); }
+
+_Use_decl_annotations_
+void WM_Handler::OnEnable(HWND hwnd, BOOL fEnable) noexcept
+{
+    ::EnableWindow(hwnd, fEnable);
+}
+
+_Use_decl_annotations_
+void WM_Handler::OnCancelMode(HWND hwnd) noexcept
+{
+    ::EnableWindow(hwnd, FALSE);
+}
+
+_Use_decl_annotations_
+HFONT WM_Handler::OnGetFont(HWND hwnd) noexcept { return GetWindowFont(hwnd); }
+
+_Use_decl_annotations_
+void WM_Handler::OnSetFont(HWND hwndCtl, HFONT hfont, BOOL fRedraw) noexcept
+{
+    SetWindowFont(hwndCtl, hfont, fRedraw);
+}
+
+_Use_decl_annotations_
+INT WM_Handler::OnGetText(HWND hwnd, int cchTextMax, LPTSTR lpszText) noexcept
+{
+    return ::GetWindowText(hwnd, lpszText, cchTextMax);
+}
+
+_Use_decl_annotations_
+void WM_Handler::OnSetText(HWND hwnd, LPCTSTR lpszText) noexcept { ::SetWindowText(hwnd, lpszText); }
+
+_Use_decl_annotations_
+INT WM_Handler::OnGetTextLength(HWND hwnd) noexcept { return ::GetWindowTextLength(hwnd); }
+
+_Use_decl_annotations_
+void WM_Handler::OnSetFocus(HWND hwnd, HWND hwndOldFocus) noexcept
+{
+    // Just for the sake of using it (compiler warning)
+    ::SetFocus(hwndOldFocus);
+    ::SetFocus(hwnd);
+}
+
+_Use_decl_annotations_
+void WM_Handler::OnKillFocus(HWND hwnd, HWND hwndNewFocus) noexcept
+{
+    this->OnSetFocus(hwndNewFocus, hwnd);
+}
+
+_Use_decl_annotations_
+void WM_Handler::OnSetRedraw(HWND hwnd, BOOL fRedraw) noexcept
+{
+    SetWindowRedraw(hwnd, fRedraw);
+}
+
+_Use_decl_annotations_
+void WM_Handler::OnShowWindow(HWND hwnd, BOOL fShow, UINT status) noexcept
+{
+    FORWARD_WM_SHOWWINDOW(hwnd, fShow, status, PostMessage);
+}
+
+_Use_decl_annotations_
+UINT WM_Handler::OnGetDlgCode(HWND hwnd, LPMSG lpmsg) noexcept
+{
+    return FORWARD_WM_GETDLGCODE(hwnd, lpmsg, SendMessage);
+}
+
+_Use_decl_annotations_
+HWND WM_Handler::OnNextDlgCtl(HWND hwnd, HWND hwndSetFocus, BOOL fNext) noexcept
+{
+    // strangely the docs said to post it
+    return FORWARD_WM_NEXTDLGCTL(hwnd, hwndSetFocus, fNext, PostMessage);
+}
+
+_Use_decl_annotations_
+void WM_Handler::OnEnterIdle(HWND hwnd, UINT source, HWND hwndSource) noexcept
+{
+    FORWARD_WM_ENTERIDLE(hwnd, source, hwndSource, ::DefWindowProc);
+}
+
+_Use_decl_annotations_
+void WM_Handler::OnNCPaint(HWND hwnd, HRGN hrgn) noexcept
+{
+    using WindowDC = std::unique_ptr<HDC__ , decltype(&::DeleteDC)>;
+    WindowDC hdc{ GetDCEx(hwnd, hrgn, DCX_WINDOW | DCX_INTERSECTRGN), ::DeleteDC };
+    ReleaseDC(hwnd, hdc.release());
```

Comment:
We did our part in not making it leak.
If Windows leaks it so be it. We can't use `ReleaseDC `as a deleter because `HDC `is a handle and doesn't have constructors.